### PR TITLE
Kore syntax

### PR DESCRIPTION
**Pull request by traiansf**
this PR does 3 things:
* removing CharLiteral
* improved Id parsing
* adding a document containing the Syntax of Kore
---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

**Comment by traiansf** [comment]
Fixes #947

File: docs/kore-syntax.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+--------------
+
+### Comments
+
+- C-style comments (non-nested)
+
+### Strings
+
+```
+StringLiteral ::= "\"" {Char}* "\""
```
Comment (virgil-serbanuta) [inline docs/kore-syntax.md]: It's fairly obvious, but, if it's not hard, I'd add a link to something describing the syntax used here, e.g. that things between quotation marks are literals, curly braces are used to identify things to which "repeaters" apply (e.g. `+*?`), and so on.

File: docs/kore-syntax.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    | "\u" {HexDigit}4
+    | "\U" {HexDigit}8
+
+HexDigit ::= ['0'..'9', 'A'..'F', 'a'..'f']
+```
+
+### KeyWords
+
+```
+Keywords = "module" | "endmodule" | "sort" | "symbol" | "alias" | "axiom"
```
Comment (virgil-serbanuta) [inline docs/kore-syntax.md]: "claim", "where", "hooked-sort", "hooked-symbol"?

File: kore/src/Kore/Parser/Lexeme.hs [hunk 3]
```diff
@@ -217,36 +229,44 @@ idCharSet :: CharSet
...
-            idToken <- objectIdRawParser KeywordsPermitted
-            return (c:idToken)
+    start <- Parser.char '@'
+    end <- genericKoreIdRawParser KeywordsPermitted
+    return (start:end)
+
+{-|'elemVarIdRawParser' extracts the string representing a
+@set-variable-identifier@. Does not consume whitespace.
+
+Always starts with @@@
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Parser/Lexeme.hs]: Should't @ be escaped, see, e.g., line 121 (documentation for `variableIdParser`)?

File: src/test/resources/test-exception-3.kore [hunk 4]
```diff
@@ -1,8 +0,0 @@
-[]
-
-module TEST-EXCEPTION-3
-
-  axiom{S}
-    'X:S []
```
Comment (virgil-serbanuta) [inline src/test/resources/test-exception-3.kore]: Do we have an equivalent test with strings?

File: kore/src/Kore/Parser/Lexeme.hs [hunk 5]
```diff
@@ -217,36 +229,44 @@ idCharSet :: CharSet
...
-objectIdRawParser :: IdKeywordParsing -> Parser String
-objectIdRawParser keywordParsing = do
-    backslash <- Parser.string "\\" <|> pure ""
-    name <- objectNonslashIdRawParser keywordParsing
-    pure (backslash ++ name)
-
-objectNonslashIdRawParser :: IdKeywordParsing -> Parser String
-objectNonslashIdRawParser =
+symbolIdRawParser :: Parser String
+symbolIdRawParser = do
```
Comment (ttuegel) [inline kore/src/Kore/Parser/Lexeme.hs]: Why not,

```.hs
symbolIdRawParser =
    ((:) <$> char '\\' <*> genericKoreIdRawParser KeywordsPermitted)
    <|> genericKoreIdRawParser KeywordsForbidden
```

(Forgiving my formatting.)

File: kore/src/Kore/Parser/Lexeme.hs [hunk 6]
```diff
@@ -97,35 +100,37 @@ sourcePosToFileLocation
...
+    name <- lexeme idRawParser
     return Id
         { getId = Text.pack name
         , idLocation = AstLocationFile pos
         }
 
+variableIdRawParser :: Parser String
+variableIdRawParser = elemVarIdRawParser <|> setVarIdRawParser
+
+{-|'variableIdParser' parses either an @element-variable-identifier@, or a
```
Comment (ttuegel) [inline kore/src/Kore/Parser/Lexeme.hs]: The documentation here and below should be updated to reflect the new syntax document.

File: docs/kore-syntax.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+### Strings
+
+```
+StringLiteral ::= "\"" {Char}* "\""
+
+Char ::= EscapeChar || AsciiChar || PrintableChar
+
+AsciiChar ::= first 128 ASCII characters
+PrintableChar ::= printable according to the C++ iswprint function
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/kore-syntax.md]: Actually their definition is pretty strange, because it depends on the current `locale` settings: I was quite surprise we do support these and I'm pretty certain that maybe we shouldn't.

File: docs/kore-syntax.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/kore-syntax.md]: You are perfectly right; Virgil also mentioned that. What is interesting is that our parsed only accepts the ones I listed :-)
I'll correct both.

File: docs/kore-syntax.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+An identifier cannot be a keyword
+
+Context-Free Syntax
+-------------------
+
+```
+Definition ::= Attributes {Module}+
+
+Attributes ::= "[" {Pattern ","}* "]"
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/kore-syntax.md]: I kind of agree with you; however, our parser currently accepts any pattern.

File: docs/kore-syntax.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+Module ::= "module" ModuleNameId {Sentences}* "endmodule" Attributes
+
+Sentence
+  ::= ImportStatement
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/kore-syntax.md]: I'm not 100% sure; if it's an internal document, OK. If we mean it as external, we might want to simplify certain names.

File: docs/kore-syntax.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    | Axiom
+    | Claim
+
+ImportStatement ::= "import" ModuleNameId Attributes
+
+SortDefinition ::= "sort" SortId SortParameters Attributes
+HookedSortDefinition ::= "hooked-sort" SortId SortParameters Attributes
+
+SymbolDefinition ::=
+    "symbol" SymbolId SortParameters "(" { Sort ","}* "}" ":" Sort Attributes
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/kore-syntax.md]: Well, in the SDF https://en.wikipedia.org/wiki/Syntax_Definition_Formalism ,  `{ Sort ","}*` is understood as not having a final comma and this notation was created especially to avoid using extra productions when defining lists.

Maybe we should clarify our syntax definition. And we could indeed define lists, if you think it would make things easier to read for newcomers.

File: kore/src/Kore/Parser/Lexeme.hs (hunk 5 above)
Comment (traiansf) [reply kore/src/Kore/Parser/Lexeme.hs]: Maybe we could rewrite more of our parsing infrastructure this way;  This was a way to clearly show that you only have 1 char lookahead. In your definition, this is harder to grasp.

File: docs/kore-syntax.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+- C-style comments (non-nested)
+
+### Strings
+
+```
+StringLiteral ::= "\"" {Char}* "\""
+
+Char ::= EscapeChar || AsciiChar || PrintableChar
+
+AsciiChar ::= first 128 ASCII characters
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/kore-syntax.md]: I just stated what our parser accepts. maybe that needs to change, too.

File: kore/src/Kore/Parser/Lexeme.hs (hunk 3 above)
Comment (traiansf) [reply kore/src/Kore/Parser/Lexeme.hs]: probably.

File: docs/kore-syntax.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply docs/kore-syntax.md]: Does our parser really accept """ as a valid string? I thought it will take the first two quotation marks as being the string and produce an error at the third.

File: docs/kore-syntax.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ttuegel) [reply docs/kore-syntax.md]: > Actually their definition is pretty strange, because it depends on the current `locale` settings

That shouldn't be the case, are you sure? The Unicode character classes are fixed. If your system is using a non-Unicode encoding, the locale settings will influence how characters are decoded from your system encoding to Unicode, but we only accept ASCII anyway. If this is really influenced by locale settings, then we need to fix the implementation.

File: kore/src/Kore/Parser/Lexeme.hs (hunk 5 above)
Comment (ttuegel) [reply kore/src/Kore/Parser/Lexeme.hs]: That's true. It's a trade-off: the current implementation shows look-ahead very clearly versus the `Alternative` definition can be made to look (almost exactly) like the syntax document. We should keep it as-is to be consistent.

File: docs/kore-syntax.md (hunk 1 above)
Comment (traiansf) [reply docs/kore-syntax.md]: I've changed it to valid BNF and added a link to wikipedia

File: src/test/resources/test-exception-3.kore (hunk 4 above)
Comment (traiansf) [reply src/test/resources/test-exception-3.kore]: good catch, we didn't.

File: docs/kore-syntax.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+[Semantics of K](docs/semantics-of-k.pdf)
+
+Any change in the KORE Parser infrastructure should be reflected here.
+
+Lexical Syntax
+--------------
+
+### Comments
+
+- C-style comments (non-nested)
```
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/kore-syntax.md]:
```.hs
skipWhitespace ::  Parser ()
skipWhitespace =
    L.space Parser.space1 lineComment blockComment
  where
    lineComment = L.skipLineComment "//"
    blockComment = L.skipBlockComment "/*" "*/"
```

So, we have line comment, and block comments are not nested as there is another function which would do that: https://hackage.haskell.org/package/megaparsec-7.0.5/docs/Text-Megaparsec-Char-Lexer.html#v:skipBlockCommentNested

File: docs/kore-syntax.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/kore-syntax.md]: I based my comment on this: https://en.cppreference.com/w/c/string/wide/iswprint
and on the fact that `Data.Char.isPrint` delegates to `iswprint`: https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.Unicode.html#isPrint

File: docs/kore-syntax.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (traiansf) [reply docs/kore-syntax.md]: I've changed everything to BNF and defined enumerations manually.

File: docs/kore-syntax.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ttuegel) [reply docs/kore-syntax.md]: It does not, it delegates to `u_iswprint` which is defined somewhere in the RTS. I'm still looking it up at the moment, but I doubt it delegates to `iswprint`, which has the wrong signature on most platforms.

[the remark this answers is not in the document]
Comment (ttuegel) [reply docs/kore-syntax.md]: `u_iswprint` is defined as this [predicate](https://gitlab.haskell.org/ghc/ghc/blob/master/libraries/base/cbits/WCsubst.c#L5054) on Unicode character classes.

File: kore/src/Kore/Parser/Lexeme.hs [hunk 13]
```diff
@@ -97,35 +100,37 @@ sourcePosToFileLocation
...
-@set-variable-identifier@.
-
-The @set-variable-@ version always starts with @\@@, while the @object-@
-one does not.
+{-| Generic identifier parser. Takes a parser for the string of the identifier
+ and returns an 'Id' annotated with position.
 -}
-idParser :: Parser Id
-idParser = do
+genericIdParser :: Parser String -> Parser Id
```
[the remark this answers is not in the document]
Comment (traiansf) [reply kore/src/Kore/Parser/Lexeme.hs]: True. I've renamed it.

File: docs/kore-syntax.md [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+Context-Free Syntax
+-------------------
+
+The definition below uses standard [BNF](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form).
+
+Nevertheless, we use following notational convention for enumerations.
+- enumeration symbols are plurals of the symbols they enumerate over.
+  For example, `<attributes>` enumerates over `<attribute>`
+- enumeration uses comma as a separator; enumeration symbols using
+  whitespace as a separator are suffices with `_`, e.g., `<sentences>`
```
Comment (andreiburdusa) [inline docs/kore-syntax.md]: I can't find the `_` in the example

File: docs/kore-syntax.md [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+
+String literals are defined in the [Lexical Syntax](#lexical-syntax) section
+below.
+
+### Enumerations
+
+```
+<modules_+>
+  ::= <module>
+    | <module> <modules+>
```
Comment (andreiburdusa) [inline docs/kore-syntax.md]: Shouldn't this be ```| <module> <modules_+>```

File: docs/kore-syntax.md [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+
+<sentences_>
+  ::= ""
+    | <sentence> <sentences>
```
Comment (andreiburdusa) [inline docs/kore-syntax.md]: Shouldn't this be ```| <sentence> <sentences_>``` ?

File: kore/test/Test/Kore/Parser/Lexeme.hs [hunk 17]
```diff
@@ -323,105 +320,6 @@ stringLiteralParserTests =
             ]
         ]
 
-charLiteralParserTests :: [TestTree]
```
Comment (ttuegel) [inline kore/test/Test/Kore/Parser/Lexeme.hs]: Where did these go?

File: src/test/resources/test-exception-3.kore.golden [hunk 18]
```diff
@@ -1,6 +1,6 @@
...
+../src/test/resources//test-exception-3.kore:6:12:
   |
-6 |     'X:S []
-  |       ^
-unexpected ':'
-expecting '''
+6 |     "X:S []
+  |            ^
+unexpected newline
+expecting '"' or printable ASCII character
```
Comment (ttuegel) [inline src/test/resources/test-exception-3.kore.golden]: Should we improve this error message? It could also be an escape sequence.

File: kore/test/Test/Kore/Parser/Lexeme.hs (hunk 17 above)
Comment (traiansf) [reply kore/test/Test/Kore/Parser/Lexeme.hs]: I just removed them, as we don't have any `CharLiteral` type; however, we have plenty of `StringLiteral` tests so we should be covered

File: src/test/resources/test-exception-3.kore.golden (hunk 18 above)
Comment (traiansf) [reply src/test/resources/test-exception-3.kore.golden]: well, it cannot be an escape sequence anymore, because it doesn't start with `\`.  The message changed because I used peekChar to look one char ahead to chose between escape chars and printable chars.

Comment (ttuegel) [reply src/test/resources/test-exception-3.kore.golden]: It could be the start of an escape sequence, i.e. it could be `\` itself.

Comment (traiansf) [reply src/test/resources/test-exception-3.kore.golden]: that's what I'm saying, that there is a check for `\` before we get here, so it couldn't be  the start of an escape sequence.

Comment (ttuegel) [reply src/test/resources/test-exception-3.kore.golden]: A `\` is permitted there by the syntax.

Comment (traiansf) [reply src/test/resources/test-exception-3.kore.golden]: I see. however, these messages are generated automatically from the "labels" we're adding, and because now we're not attempting `\` anymore in this context, it got "automagically" removed. Not sure how to put it back.

Comment (ttuegel) [reply src/test/resources/test-exception-3.kore.golden]: That makes sense. I wouldn't worry about such a corner case right now.
